core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/TransmissionNetworkOutput.java

```diff
@@ -33,294 +34,188 @@
+		Preconditions.checkArgument(!Strings.isNullOrEmpty(serverUri), "serverUri should be a valid non-null value");
+		Preconditions.checkNotNull(transmissionPolicyManager,
+				"transmissionPolicyManager should be a valid non-null value");
+
+		this.serverUri = serverUri;
+
+		httpClient = ApacheSenderFactory.INSTANCE.create();
+		this.transmissionPolicyManager = transmissionPolicyManager;
+		stopped = false;
+		
+	}
+
+	public void setTransmissionDispatcher(TransmissionDispatcher transmissionDispatcher) {
+		this.transmissionDispatcher = transmissionDispatcher;
+	}
+	
+	/**
+	 * Stops all threads from sending data.
+	 * 
+	 * @param timeout
+	 *            The timeout to wait, which is not relevant here.
+	 * @param timeUnit
+	 *            The time unit, which is not relevant in this method.
+	 */
+	@Override
+	public synchronized void stop(long timeout, TimeUnit timeUnit) {
+		if (stopped) {
+			return;
+		}
+
+		httpClient.close();
+		stopped = true;
+	}
+
+	/**
+	 * Tries to send a
+	 * {@link com.microsoft.applicationinsights.internal.channel.common.Transmission}
+	 * The thread that calls that method might be suspended if there is a throttling
+	 * issues, in any case the thread that enters this method is responsive for
+	 * 'stop' request that might be issued by the application.
+	 * 
+	 * @param transmission
+	 *            The data to send
+	 * @return True when done.
+	 */
+	@Override
+	public boolean send(Transmission transmission) {
+		if (!stopped) {
+			// If we're not stopped but in a blocked state then fail to second TransmissionOutput
+			if (transmissionPolicyManager.getTransmissionPolicyState()
+					.getCurrentState() != TransmissionPolicy.UNBLOCKED) {
+				return false;
+			}
+			
+			HttpResponse response = null;
+			HttpPost request = null;
+			int code = 0;
+			String respString = null;
+			Throwable ex = null;
+			Header retryAfterHeader = null;
+			try {
+				// POST the transmission data to the endpoint
+				request = createTransmissionPostRequest(transmission);
+				httpClient.enhanceRequest(request);
+				response = httpClient.sendPostRequest(request);
+				HttpEntity respEntity = response.getEntity();
+				code = response.getStatusLine().getStatusCode();
+				respString = EntityUtils.toString(respEntity);
+				retryAfterHeader = response.getFirstHeader(RESPONSE_THROTTLING_HEADER);
+				
+				// After the third time through this dispatcher we should reset the counter and then fail to second TransmissionOutput
+				if (code > HttpStatus.SC_PARTIAL_CONTENT && transmission.getNumberOfSends() >= 3) {
+					transmission.setNumberOfSends(0);
+					return false;
+				} else if (code == HttpStatus.SC_OK) {
+					// If we've completed then clear the back off flags as the channel does not need to be throttled
+					transmissionPolicyManager.clearBackoff();
+				}
+				return true;
+					
+			} catch (ConnectionPoolTimeoutException e) {
+				ex = e;
```

Comment:
All this places we need to dump the stack trace. Please use ExceptionUtils to do so and pass it in logger.